include/jemalloc/internal/ph.h

```diff
@@ -414,6 +414,73 @@ ph_remove(ph_t *ph, void *phn, size_t offset, ph_cmp_t cmp) {
 	}
 }
 
+JEMALLOC_ALWAYS_INLINE void
+ph_queue_push(void *phn, void **bfs_queue, uint16_t *rear, size_t *queue_size,
+    size_t max_queue_size) {
+	assert(*queue_size < max_queue_size);
+	bfs_queue[*rear] = phn;
+	*rear = (*rear + 1) % max_queue_size;
+	(*queue_size) ++;
+}
+
+JEMALLOC_ALWAYS_INLINE void *
+ph_queue_pop(void **bfs_queue, uint16_t *front, size_t *queue_size,
+    size_t max_queue_size) {
+	assert(*queue_size > 0);
+	void *ret = bfs_queue[*front];
+	*front = (*front + 1) % max_queue_size;
+	(*queue_size) --;
+	return ret;
+}
+
+/*
+ * The two functions below offer a solution to enumerate the pairing heap.
+ * Whe enumerating, always call ph_enumerate_prepare first to prepare the queue
+ * needed for BFS.  Next, call ph_enumerate_next to get the next element in
+ * the enumeration.  When enumeration ends, ph_enumerate_next returns NULL and
+ * should not be called again.  Enumeration ends when all elements in the heap
+ * has been enumerated or the number of visited elements exceed
+ * max_visit_num.
+ */
+JEMALLOC_ALWAYS_INLINE void
+ph_enumerate_prepare(ph_t *ph, void **bfs_queue, uint16_t *front,
+    uint16_t *rear, size_t *queue_size, size_t *visited_num,
```

Comment:
Since we pass several stack args here, better make it a struct that contains all these. That way you can also add initializer or access helpers and include asserts in them.

Reply:
Grouped them into a `ph_enumerate_helper_t` where all variables including the array `bfs_queue` is included. I include the array in the helper to prevent users from declaring any extra stack variables besides helper. Except for the array, other variables are put into `ph_enumerate_vars_t`, which serves the purposes suggested here. 

Also, I changed all things in vars to `uint16_t` because I don't think the enumeration should allow maximum visited nodes to reach 60K: it is going to damage the perf too much.